journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java

```diff
@@ -105,72 +133,97 @@ public void append(final JournalRecord record) {
-    final var checksum = checksumGenerator.compute(record.data());
-    if (checksum != record.checksum()) {
-      throw new InvalidChecksum("Checksum invalid for record " + record);
+    if (record.checksum() != checksum) {
+      buffer.position(startPosition);
+      throw new InvalidChecksum(
+          String.format("Failed to append record %s. Checksum does not match", record));
     }
-    writeInternal(buffer, record);
-    return record;
+
+    final JournalRecordMetadata metadata =
+        writeMetadata(buffer, writeBuffer, startPosition, metadataLength, recordLength, checksum);
+
+    updateLastWrittenEntry(metadata, indexedRecord, startPosition);
+
+    buffer.position(startPosition + metadataLength + recordLength);
   }
 
-  private void writeInternal(final ByteBuffer buffer, final JournalRecord recordToWrite) {
-    final int recordStartPosition = buffer.position();
-    buffer.mark();
-    if (recordStartPosition + Integer.BYTES > buffer.limit()) {
-      throw new BufferOverflowException();
-    }
+  private void updateLastWrittenEntry(
+      final JournalRecordMetadata metadata,
+      final JournalIndexedRecord recordRead,
+      final int startPosition) {
+    lastEntry = new PersistedJournalRecord(metadata, recordRead);
+    index.index(lastEntry, startPosition);
+  }
 
-    buffer.position(recordStartPosition + Integer.BYTES);
-    try {
-      serializer.write(recordToWrite, buffer);
-    } catch (final BufferOverflowException e) {
-      buffer.reset();
-      throw e;
-    }
+  private JournalRecordMetadata writeMetadata(
+      final ByteBuffer buffer,
+      final MutableDirectBuffer writeBuffer,
+      final int startPosition,
+      final int metadataLength,
+      final int recordLength,
+      final long checksum) {
+    writeBuffer.wrap(buffer, startPosition, metadataLength);
+    final JournalRecordMetadata recordMetadata =
+        new JournalRecordMetadataImpl(checksum, recordLength);
+    serializer.write(recordMetadata, writeBuffer);
+    return recordMetadata;
+  }
 
-    final int length = buffer.position() - (recordStartPosition + Integer.BYTES);
+  private int writeRecord(
+      final ByteBuffer buffer,
+      final MutableDirectBuffer writeBuffer,
+      final int offset,
+      final JournalIndexedRecord indexedRecord) {
+    writeBuffer.wrap(buffer, offset, buffer.limit() - offset);
+    final var recordLength = serializer.write(indexedRecord, writeBuffer);
+    return recordLength;
+  }
 
-    // If the entry length exceeds the maximum entry size then throw an exception.
-    if (length > maxEntrySize) {
-      // Just reset the buffer. There's no need to zero the bytes since we haven't written the
-      // length or checksum.
-      buffer.reset();
+  private void checkCanWrite(final ByteBuffer buffer, final JournalIndexedRecord indexedRecord) {
+    final var recordOffset = serializer.getMetadataLength();
+
+    final int estimatedRecordLength =
```

Comment:
If we're estimating, then we probably want to estimate greater than or equal to the real size right? Is it what we're doing now? If we don't, then we could estimate something smaller than the length and fail to write it right? But I guess at worse we just would write in the next segment, so...

Reply:
KryoSerializer throws BufferOverflowException if there is not enough space to write in this segment. This estimated length is mostly only used to check with the maxEntrySize which I guess we want to remove it. So this is only a temporary state.